README.adoc

```diff
@@ -0,0 +1,180 @@
+
+A Section can be implemented as a Jinja template or as a Python function.
+In either case, it is invoked through Scribble's "Section" call.
+```
+Section("module name", scope, **context)
+```
+where:
+
+[horizontal]
+name:: is the python or jinja module name.
+scope:: is a subtree (element) of the object model.
+context:: are additional named values passed to the section.
+
+When a section is written as a Jinja template, all the
+named arguments are available as Jinja template variables.
+
+
+=== "Overview" Paragraph
+
+The overview is a simple paragraph describing the block.
+
+It is usually written as a single Jinja template.
+It doesn't need a lot of detail, but it isn't entirely static either.
+For example, if there is more than one instance of the device, then this
+section would mention how many there are.
+
+As an example, in the file `Overview.jinja2` the following line would describe how many parallel ports are in the design.
+
+----
+The {{ product_name }} contains {{ devices | length | english_number }} Parallel I/O ports.
+----
+
+which would be displayed as:
+
+----
+The PIO Test Socket contains five Parallel I/O ports.
+----
+
+=== "Programming" Chapter
+This chapter describes the device or block and how to use it.
+For some devices, the Section can get quite large and include a hierarchy
+of sections and subsections. For simpler blocks, it might be a single Jinja
+template file named `Programming.jinja2`.
+
+
+=== "Hardware Interface" Chapter
+The hardware interface chapter decribes how the block interfaces with the rest of the design. It would
+describe the input and output signals, how they are clocked, and what interrupts the device generates.
+If it contains SRAMs, it might describe the SRAMs and how they are used.
+
+Again depending on the block, this chapter could invoke a hierarchy of sections and subsections. For simpler blocks,
+it might consist of a single Jinja template file named `HardwareInterface.jinja2`.
+
+== Appendix: Jinja Templates
+
+=== Overview
+Jinja templates consist of text interspersed with Jinja control statements.
+For a good introduction to Jinja, see the online https://jinja.palletsprojects.com/en/2.10.x/[Jinja Documentation].
+
+When Scribble invokes a Jinja template, it passes a number of predefined variables and template functions to the template.
+These variables and functions are described below.
+
+==== Predefined Variables for Onboarding New Blocks
+When documenting new blocks, each of the three sections will be passed the following predefined variables.
+
+
+scope:: The primary element being described. Typically this will be
+        based on the object model schema for the block.
+
+devices:: If there are multiple instances of the block, then
+  `devices` contains all of them.
+   Typically, `scope` is used to describe what all the devices
+   have in common, while `devices` is used to iterate through
+   the instances describing the unique values of each.
```

Comment:
What exactly is getting passed in as `scope` to the user-provided sections? This seems to imply that in certain contexts, `scope` may have different values, but the person onboarding a block has no control over it, so it's not clear what affects what is actually getting passed in.